# 01. Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Assigning variable to folder path

path = r'C:\Users\Amy Yip\Desktop\Career Foundry Data Analytics\Achievement 4 Python\05-02-2022 Instacart Basket Analysis\02 Data'

#### 1. Merged your prepared Instacart orders data with the new orders_products_prior dataframe per the instructions in this 4.6.

In [3]:
# Import new dataset (order_products_prior.csv) from this 4.6

df_ords_prior = pd.read_csv(os.path.join(path, 'Original Data', 'order_products_prior.csv'), index_col = False)

In [4]:
# Import the latest orders dataset (orders_checked_Ex4.5.csv)

df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_checked_Ex4.5.csv'), index_col = False)

In [5]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [6]:
df_ords.head()

Unnamed: 0  Unnamed: 0.1  order_id  user_id  order_numbers_users_made  \
0           0             0   2539329        1                         1   
1           1             1   2398795        1                         2   
2           2             2    473747        1                         3   
3           3             3   2254736        1                         4   
4           4             4    431534        1                         5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   2                  8                     NaN   
1                   3                  7                    15.0   
2                   3                 12                    21.0   
3                   4                  7                    29.0   
4                   4                 15                    28.0   

   prior_orders_made  
0               True  
1              False  
2              False  
3              False  
4              False

In [7]:
# Exclude the import of the first 2 columns

vars_list = ['order_id', 'user_id', 'order_numbers_users_made', 'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order', 'prior_orders_made']

In [8]:
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_checked_Ex4.5.csv'), usecols = vars_list)

In [9]:
df_ords.head()

order_id  user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2398795        1                         2                   3   
2    473747        1                         3                   3   
3   2254736        1                         4                   4   
4    431534        1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  
0                  8                     NaN               True  
1                  7                    15.0              False  
2                 12                    21.0              False  
3                  7                    29.0              False  
4                 15                    28.0              False

In [10]:
# Check the dimensions of the imported dataframe

df_ords_prior.shape

(32434489, 4)

In [11]:
df_ords.shape

(3421083, 7)

In [12]:
# Merge the dataframe with your df_ords dataframe using default how='inner'

df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [13]:
df_merged_large.head()

order_id  user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2539329        1                         1                   2   
2   2539329        1                         1                   2   
3   2539329        1                         1                   2   
4   2539329        1                         1                   2   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  8                     NaN               True       14084   
2                  8                     NaN               True       12427   
3                  8                     NaN               True       26088   
4                  8                     NaN               True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [14]:
# Determine whether there is a full match: (not quite the right method since we used inner join above)

df_merged_large['_merge'].value_counts()

both          32434489
right_only           0
left_only            0
Name: _merge, dtype: int64

In [15]:
df_merged_large.shape

(32434489, 11)

Always double-check your merge rates using an outer join as well especially when you’re exploring new data and performing test merges.

In [16]:
# Check whether there really is a full match:
# Merge the dataframe with your df_ords dataframe using default how='outer'

df_merged_large_outer = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True, how = 'outer')

In [17]:
df_merged_large_outer['_merge'].value_counts()

both          32434489
left_only       206209
right_only           0
Name: _merge, dtype: int64

After the merge, there are a total of 32,434,489 rows of information found in both input data sets and 206,209 rows of information that only appears in the left dataframe.

Value of “left_only” indicates observations whose merge key only appears in the left dataframe.

Value of “both” indicates observation’s merge key is found in both dataframe.

In [18]:
df_merged_large_outer.shape

(32640698, 11)

Note: For your Instacart project, you’ll only be working with data sets that have a full merge rate, so you won’t need to worry about this or apply any changes to the merge you just completed (using how = 'inner').

So will only be using the innerjoin result.

Also looking at the orders dataframe, it contains order_id and user_id but it doesn't contain product information. On the other hand, orders_products_prior dataframe contains detail about the order_id and product_id, product information, therefore only inner join is relevant.

In [23]:
# Export data to csv (from outer join)

df_merged_large_outer.to_csv(os.path.join(path, 'Prepared Data', 'orders_products_combined_outer.csv'))

In [24]:
# Export data to pkl (from outer join)

df_merged_large_outer.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_combined_outer.pkl'))

In [21]:
# Export data to csv (from inner join)

df_merged_large.to_csv(os.path.join(path, 'Prepared Data', 'orders_products_combined_inner.csv'))

In [22]:
# Export data to pkl (from inner join)

df_merged_large.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_combined_inner.pkl'))